lizard_auth_server/forms.py

```diff
@@ -283,10 +287,80 @@ def clean(self):
             if organisation_role.role.code == BILLING_ROLE and
             organisation_role.role.portal.name == THREEDI_PORTAL])
         if num_threedi_billing_roles == 0:
-            self._errors['roles'] = self.error_class([_('required 3Di billing role is missing')])
+            self._errors['roles'] = self.error_class([
+                _('required 3Di billing role is missing')])
         if num_threedi_billing_roles >= 2:
             self._errors['roles'] = self.error_class([
                 _('Only one 3Di billing role is allowed ({} found)').format(
                     num_threedi_billing_roles)])
         print(self._errors)
         return cleaned_data
+
+
+class JWTAuthenticationForm(AuthenticationForm):
+    """An authentication form for requesting JSON Web Tokens.
+
+    This form resembles the standard authentication form. It has one extra
+    hidden `portal` field for passing the public key around of the portal
+    that the user wants access to.
+
+    """
+    portal = forms.CharField(widget=forms.HiddenInput())
+
+    def clean_portal(self):
+        """Check if the sso_key exists.
+
+        Since the `portal` field is hidden, any field errors are hidden too.
+        For that reason, if an error occurs, it is also registered as a
+        form-wide error, so it will be visible to end users (form-
+        wide errors are displayed at the top of the page).
+
+        It's unlikely that this check does not pass, because validation has
+        also been done on the GET request - unless somebody is tampering
+        with the form data...
+
+        Returns:
+            Portal: an instance of lizard_auth_server.models.Portal.
+
+        Raises:
+            ValidationError: If no portal with this sso_key exists.
+
+        """
+        data = self.cleaned_data['portal']
+        try:
+            return Portal.objects.get(sso_key=data)
+        except Portal.DoesNotExist:
+            message = _('Invalid `portal` form field.')
+            # Next line is a bit ugly. Django 1.7+ has an `add_error` method:
+            self.errors[NON_FIELD_ERRORS] = self.error_class([message])
+            # This will remove the `portal` field from cleaned_data:
+            raise ValidationError(message)
+
+    def clean(self):
+        """Check if the user has access to the portal.
+
+        Returns:
+            dict: the validated form data.
+
+        Raises:
+            ValidationError: If the user has no access to the portal.
+
+        """
+        cleaned_data = super(JWTAuthenticationForm, self).clean()
+
+        if self.errors:
+            return cleaned_data
+
+        # At this point, the user has succesfully been authenticated.
+        # Having an SSO account, however, does not grant you access
+        # to all portals managed by the SSO server, so an extra
+        # check is needed.
+
+        portal = cleaned_data.get('portal')
+        user = self.get_user()
+
+        if not UserProfile.has_profile(user, portal):
```

Comment:
Die bestaat niet, misschien bedoel je deze:

    user.has_access(portal)

Die had ik niet gezien. De huidige implementatie zegt echter:

    if self.user.is_staff:
        # staff can access any site
        return True

Ik vraag me af of dat is wat we willen?